File: dart/test/kanji_old_new_test.dart
import 'package:test/test.dart';
import 'package:yosina/yosina.dart';

void main() {
  group('KanjiOldNew', () {
    late Transliterator transliterator;

    setUp(() {
      transliterator = Transliterator.withList(['kanjiOldNew']);
    });

    group('Basic functionality without IVS', () {
      test('preserves kanji without IVS', () {
        final input = '檜';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('檜')); // No conversion without IVS
      });

      test('preserves multiple kanji without IVS', () {
        final input = '檜辻';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('檜辻')); // No conversion without IVS
      });

      test('preserves mixed content without IVS', () {
        final input = '新檜字';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('新檜字')); // No conversion without IVS
      });

      test('preserves sentence context without IVS', () {
        final input = '檜の木材を使用しています';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('檜の木材を使用しています')); // No conversion without IVS
      });

      test('no change for regular kanji', () {
        final input = '普通の漢字です';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('普通の漢字です'));
      });

      test('empty string', () {
        final input = '';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals(''));
      });

      test('mixed content with latin', () {
        final input = 'Hello 檜 World';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('Hello 檜 World')); // No conversion without IVS
      });

      test('preserve other characters', () {
        final input = '🎌檜🌸新🗾';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('🎌檜🌸新🗾')); // No conversion without IVS
      });

      test('hiragana and katakana preservation', () {
        final input = 'ひらがな檜カタカナ';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('ひらがな檜カタカナ')); // No conversion without IVS
      });

      test('numbers and punctuation', () {
        final input = '檜123！？';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('檜123！？')); // No conversion without IVS
      });
    });

    group('IVS (Ideographic Variation Selector) handling', () {
      test('kanji with IVS VS17', () {
        final input = '檜\u{E0100}';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('桧\u{E0100}'));
      });

      test('kanji with IVS VS18 to VS17', () {
        final input = '辻\u{E0101}';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('辻\u{E0100}'));
      });

      test('mixed content with IVS', () {
        final input = 'Hello 檜\u{E0100} World';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('Hello 桧\u{E0100} World'));
      });

      test('multiple kanji with IVS', () {
        final input = '檜\u{E0100}辻\u{E0101}';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('桧\u{E0100}辻\u{E0100}'));
      });

      test('sentence with IVS kanji', () {
        final input = '檜\u{E0100}の木は美しい';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('桧\u{E0100}の木は美しい'));
      });
    });

    group('Edge cases without IVS', () {
      test('kanji at string start without IVS', () {
        final input = '檜が好き';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('檜が好き')); // No conversion without IVS
      });

      test('kanji at string end without IVS', () {
        final input = 'これは檜';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('これは檜')); // No conversion without IVS
      });

      test('consecutive kanji without IVS', () {
        final input = '檜檜檜';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('檜檜檜')); // No conversion without IVS
      });

      test('kanji between non-cjk without IVS', () {
        final input = 'A檜B檜C';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('A檜B檜C')); // No conversion without IVS
      });

      test('newlines and whitespace without IVS', () {
        final input = '檜\n木\t材';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('檜\n木\t材')); // No conversion without IVS
      });

      test('unicode surrogates preservation without IVS', () {
        final input = '𝓗檜𝓮𝓵𝓵𝓸';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('𝓗檜𝓮𝓵𝓵𝓸')); // No conversion without IVS
      });

      test('mixed scripts without IVS', () {
        final input = 'English檜日本語카타나hiraganaひらがな';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result,
            equals('English檜日本語카타나hiraganaひらがな')); // No conversion without IVS
      });

      test('punctuation around kanji without IVS', () {
        final input = '(檜)、[檜]、「檜」';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('(檜)、[檜]、「檜」')); // No conversion without IVS
      });
    });

    group('Common old-new kanji conversions with IVS', () {
      test('converts 廣 to 広 with IVS', () {
        final input = '廣\u{E0100}島';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('広\u{E0100}島'));
      });

      test('converts 國 to 国 with IVS', () {
        final input = '國\u{E0100}語';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('国\u{E0100}語'));
      });

      test('converts multiple old kanji with IVS in a sentence', () {
        final input = '古い漢字：廣\u{E0100}島、檜\u{E0100}、國\u{E0100}';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('古い漢字：広\u{E0100}島、桧\u{E0100}、国\u{E0100}'));
      });

      test('preserves kanji without IVS', () {
        final input = '廣島、檜、國';
        final result =
            Chars.charsToString(transliterator(Chars.fromString(input)));
        expect(result, equals('廣島、檜、國')); // No conversion without IVS
      });
    });
  });
}

File: go/transliterators/kanji_old_new/kanji_old_new_test.go
package kanji_old_new

import (
	"testing"

	"github.com/stretchr/testify/assert"

	yosina "github.com/yosina-lib/yosina/go"
)

func TestKanjiOldNew(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "preserve unmapped kanji",
			input:    "檜",
			expected: "檜",
		},
		{
			name:     "preserve multiple kanji",
			input:    "檜辻",
			expected: "檜辻",
		},
		{
			name:     "preserve mixed content",
			input:    "新檜字",
			expected: "新檜字",
		},
		{
			name:     "preserve sentence context",
			input:    "檜の木材を使用しています",
			expected: "檜の木材を使用しています",
		},
		{
			name:     "no change for regular kanji",
			input:    "普通の漢字です",
			expected: "普通の漢字です",
		},
		{
			name:     "empty string",
			input:    "",
			expected: "",
		},
		{
			name:     "mixed content with latin",
			input:    "Hello 檜 World",
			expected: "Hello 檜 World",
		},
		{
			name:     "preserve other characters",
			input:    "🎌檜🌸新🗾",
			expected: "🎌檜🌸新🗾",
		},
		{
			name:     "hiragana and katakana preservation",
			input:    "ひらがな檜カタカナ",
			expected: "ひらがな檜カタカナ",
		},
		{
			name:     "numbers and punctuation",
			input:    "檜123！？",
			expected: "檜123！？",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := yosina.StringFromChars(
				Transliterate(yosina.NewCharIteratorFromSlice(yosina.BuildCharArray(tt.input))),
			)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestKanjiOldNewWithIVS(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "kanji with IVS selector preserved",
			input:    "檜\U000E0100",
			expected: "檜\U000E0100",
		},
		{
			name:     "kanji with different IVS selector preserved",
			input:    "辻\U000E0101",
			expected: "辻\U000E0101",
		},
		{
			name:     "mixed content with IVS preserved",
			input:    "Hello 檜\U000E0100 World",
			expected: "Hello 檜\U000E0100 World",
		},
		{
			name:     "multiple kanji with IVS preserved",
			input:    "檜\U000E0100辻\U000E0101",
			expected: "檜\U000E0100辻\U000E0101",
		},
		{
			name:     "sentence with IVS kanji preserved",
			input:    "檜\U000E0100の木は美しい",
			expected: "檜\U000E0100の木は美しい",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := yosina.StringFromChars(
				Transliterate(yosina.NewCharIteratorFromSlice(yosina.BuildCharArray(tt.input))),
			)
			assert.Equal(t, tt.expected, result)
		})
	}
}

func TestKanjiOldNewEdgeCases(t *testing.T) {
	tests := []struct {
		name     string
		input    string
		expected string
	}{
		{
			name:     "kanji at string start",
			input:    "檜が好き",
			expected: "檜が好き",
		},
		{
			name:     "kanji at string end",
			input:    "これは檜",
			expected: "これは檜",
		},
		{
			name:     "consecutive kanji",
			input:    "檜檜檜",
			expected: "檜檜檜",
		},
		{
			name:     "kanji between non-cjk",
			input:    "A檜B檜C",
			expected: "A檜B檜C",
		},
		{
			name:     "newlines and whitespace",
			input:    "檜\n木\t材",
			expected: "檜\n木\t材",
		},
		{
			name:     "unicode surrogates preservation",
			input:    "𝓗檜𝓮𝓵𝓵𝓸",
			expected: "𝓗檜𝓮𝓵𝓵𝓸",
		},
		{
			name:     "mixed scripts",
			input:    "English檜日本語카타나hiraganaひらがな",
			expected: "English檜日本語카타나hiraganaひらがな",
		},
		{
			name:     "punctuation around kanji",
			input:    "(檜)、[檜]、「檜」",
			expected: "(檜)、[檜]、「檜」",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := yosina.StringFromChars(
				Transliterate(yosina.NewCharIteratorFromSlice(yosina.BuildCharArray(tt.input))),
			)
			assert.Equal(t, tt.expected, result)
		})
	}
}
